### PR TITLE
Verkle trie refactor

### PR DESCRIPTION
**Pull request by tanishqjasoria**

### CODE REVIEW

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 1]
```diff
@@ -109,56 +112,77 @@ public void SetValue(Span<byte> rawKey, byte[] value)
...
     {
-         // is it guaranteed that the its a 12 length byte array initialized with zeros?
-         byte[] addressPadding = new byte[12] ;
-         IEnumerable<byte> treeKeyPrecursor = addressPadding.Concat(address.Bytes);
-         treeKeyPrecursor = treeKeyPrecursor.Concat(treeIndex.ToBigEndian());
-         return Sha2.Compute(treeKeyPrecursor.ToArray());
+         byte[] keyPrefix = new byte[64];
+         Array.Copy(address.Bytes, 0, keyPrefix, 12, 20);
+         treeIndex.ToBigEndian(keyPrefix.AsSpan(32));
+         return Sha2.Compute(keyPrefix);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:131]: If we could rewrite Sha2.Compute to take a Span<byte> and we probably can, we could stackallok keyPrefix, making it faster and eliminating GC.

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 2]
```diff
@@ -220,15 +244,15 @@ public byte[] GetTreeKeyForStorageSlot(Address address, UInt256 storageKey)
         
     public void SetCode(Address address, byte[] code)
     {
-        byte[][] chunkifiedCode = chunkifyCode(code);
+        byte[][] chunkifiedCode = ChunkifyCode(code);
         byte[] chunkKey;
         for (int i = 0; i < chunkifiedCode.Length; i++)
         {
             chunkKey = GetTreeKeyForCodeChunk(address, (UInt256)i);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:251]: Does key for chunk need Uint256 arithmetic? (maybe yes)

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 3]
```diff
@@ -220,15 +244,15 @@ public byte[] GetTreeKeyForStorageSlot(Address address, UInt256 storageKey)
...
             SetValue(chunkKey, chunkifiedCode[i]);
         }
     }
-    private byte[][] chunkifyCode(byte[] code)
+    private byte[][] ChunkifyCode(byte[] code)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:255]: This needs rework not to allocate array of arrays. Maybe method like:
Span<byte> TryGetNextChunk(byte[] code, ref currentPosition)

And usage:
int currentPosition = 0;
do
{
  var chunk = TryGetNextChunk(code, ref currentPosition);
  chunkKey = GetTreeKeyForCodeChunk(address, (UInt256)i);
  SetValue(chunkKey, chunk);
}
while (currentPosition < code.Length)

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
Are test not meant to pass?

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 4]
```diff
@@ -79,6 +79,7 @@ public VerkleTree(ILogManager? logManager)
         _verkleTrieObj = RustVerkleLib.VerkleTrieNew();
             
         _logger = logManager?.GetClassLogger<VerkleTree>() ?? throw new ArgumentNullException(nameof(logManager));
+        _logger.Info(_verkleTrieObj.ToString());
```
[the remark this answers is not in the document]
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Trie/VerkleTree.cs:82]: yes, this is temporary

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs (hunk 2 above)
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Trie/VerkleTree.cs:251]: yes, that's necessary

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 5]
```diff
@@ -267,22 +303,25 @@ private byte[][] chunkifyCode(byte[] code)
 
             pos += pushLength;
         }
-            
-        int chunkCount = (chunkifyableCode.Length + 31) / 32;
-        byte[][] chunks = new byte[chunkCount][];
+        
+        int chunkCount = (code.Length + 30) / 31; 
+        Span<byte> chunkifyableCode = new byte[chunkCount * 32];
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:308]: Still allocates a lot

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 6]
```diff
@@ -267,22 +303,25 @@ private byte[][] chunkifyCode(byte[] code)
...
-        {
-            chunks[i] = new byte[32];
-            chunks[i][0] = (byte)Math.Min(bytesToExecData[pos], 31);
-            Buffer.BlockCopy(chunkifyableCode, pos, chunks[i], 1, 31);
-            pos = pos + 31;
+        {   
+            codeCursor[0] = (byte)Math.Min(bytesToExecData[pos], 31);
+            codeCursor = codeCursor.Slice(1);
+            
+            code.Slice(pos, code.Length - pos < 31? code.Length - pos: 31).CopyTo(codeCursor);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:317]: still does a lot of copying

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 7]
```diff
@@ -220,33 +256,33 @@ public byte[] GetTreeKeyForStorageSlot(Address address, UInt256 storageKey)
...
         const int PUSH32 = PUSH_OFFSET + 32;
-            
-        // To ensure that the code can be split into chunks of 31 bytes
-        byte[] chunkifyableCode = new byte[code.Length + 31 - code.Length % 31];
-        Buffer.BlockCopy(code, 0, chunkifyableCode, 0, code.Length);
-
-        int[] bytesToExecData  = new int[chunkifyableCode.Length];
+        
+        var codeLength = code.Length;
+        int[] bytesToExecData  = new int[codeLength];
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:277]: allocates quite a lot

File: src/Nethermind/Nethermind.Trie/VerkleTree.cs [hunk 8]
```diff
@@ -265,12 +265,22 @@ public void SetCode(Address address, byte[] code)
...
         while (chunkEnumerator.TryGetNextChunk(out byte[] chunk))
         {
-            byte[] chunkKey = GetTreeKeyForCodeChunk(address, (UInt256)i);
+            // byte[] chunkKey = GetTreeKeyForCodeChunk(address, (UInt256)i);
+            
+            // find tree index and the sub index in verkle tree for the code chunk
+            UInt256 chunkOffset = CodeOffset + i;
+            UInt256 treeIndex = chunkOffset / VerkleNodeWidth;
+            UInt256.Mod(chunkOffset, VerkleNodeWidth, out UInt256 subIndex);
+            subIndex.ToBigEndian(subIndexBytes);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/VerkleTree.cs:312]: This probably could/should be in separate method, like FillTreeKeyForCodeChunk taking Span

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**